src/hotspot/share/prims/jvm.cpp

```diff
@@ -3704,6 +3705,69 @@ JVM_ENTRY(void, JVM_LogLambdaFormInvoker(JNIEnv *env, jstring line))
 #endif // INCLUDE_CDS
 JVM_END
 
+#if INCLUDE_CDS
+class DumpClassListCLDClosure : public CLDClosure {
+  fileStream *_stream;
+public:
+  DumpClassListCLDClosure(fileStream* f) : CLDClosure() { _stream = f; }
+  void do_cld(ClassLoaderData* cld) {
+    for (Klass* klass = cld->klasses(); klass != NULL; klass = klass->next_link()) {
+      if (klass->is_instance_klass()) {
+        InstanceKlass* ik = InstanceKlass::cast(klass);
+        if (ik->is_shareable()) {
+          _stream->print_cr("%s", ik->name()->as_C_string());
+        }
+      }
+    }
+  }
+};
+#endif
+
+JVM_ENTRY(void, JVM_DumpClassListToFile(JNIEnv *env, jstring listFileName))
+#if INCLUDE_CDS
+  ResourceMark rm(THREAD);
+  Handle file_handle(THREAD, JNIHandles::resolve_non_null(listFileName));
+  char* file_name  = java_lang_String::as_utf8_string(file_handle());
+  fileStream stream(file_name, "w");
+  if (stream.is_open()) {
+    MutexLocker lock(ClassLoaderDataGraph_lock);
+    DumpClassListCLDClosure collect_classes(&stream);
+    ClassLoaderDataGraph::loaded_cld_do(&collect_classes);
+  } else {
+    THROW_MSG(vmSymbols::java_io_IOException(), "Failed to open file");
+  }
+#endif // INCLUDE_CDS
+JVM_END
+
+JVM_ENTRY(void, JVM_DumpDynamicArchive(JNIEnv *env, jstring archiveName))
+#if INCLUDE_CDS
+  assert(UseSharedSpaces && RecordDynamicDumpInfo, "already checked in arguments.cpp?");
+  if (DynamicArchive::has_been_dumped_once()) {
+    THROW_MSG(vmSymbols::java_lang_RuntimeException(),
+        "Dynamic dump has been done, and should only be done once");
+  } else {
+    // prevent multiple dumps.
+    DynamicArchive::set_has_been_dumped_once();
+  }
+  assert(ArchiveClassesAtExit == nullptr, "already checked in arguments.cpp?");
+  Handle file_handle(THREAD, JNIHandles::resolve_non_null(archiveName));
+  char* archive_name  = java_lang_String::as_utf8_string(file_handle());
+  ArchiveClassesAtExit = archive_name;
+  if (Arguments::init_shared_archive_paths()) {
+    DynamicArchive::dump();
+  } else {
+    THROW_MSG(vmSymbols::java_lang_RuntimeException(),
```

Comment:
Need to set ArchiveClassesAtExit to NULL before throwing the exception, since dynamic dump may not work anymore after the failure.